### PR TITLE
#1783 Less logs for circuit breakers (Polly exceptions)

### PR DESCRIPTION
**Pull request by RaynaldM**
## Proposed Changes
Remove try/catch in PollyPoliciesDelegatingHandler
Add a more generic AddPolly<T> to be able to use a specific PollyQoSProvider

### CODE REVIEW

File: src/Ocelot.Provider.Polly/OcelotBuilderExtensions.cs
Comment (raman-m) [inline src/Ocelot.Provider.Polly/OcelotBuilderExtensions.cs:18]: Too long line of code! Formatting issue.
It is better to put each parameter on new line.
`where` on new line too.

File: src/Ocelot.Provider.Polly/PollyPoliciesDelegatingHandler.cs
Comment (raman-m) [inline src/Ocelot.Provider.Polly/PollyPoliciesDelegatingHandler.cs:2]: Please, don't shuffle usings!
Do you use Visual Studio or other IDE? Which one?

File: src/Ocelot.Provider.Polly/PollyPoliciesDelegatingHandler.cs
Comment (raman-m) [inline src/Ocelot.Provider.Polly/PollyPoliciesDelegatingHandler.cs:43]: You need to complete these empty XML-docs lines.
Otherwise just remove empty tags.

**Review by raman-m — CHANGES_REQUESTED** [review]
Failed 4 tests!
And some minor issues

**Comment by RaynaldM** [comment]
For "should_be_invalid_re_route_using_downstream_http_version" UT, I noticed that the message changes depending on the OS language.
image
So, I have changed it, and I fix ThenTheErrorsContains method so that it makes "Contains" and not "Equal".

File: src/Ocelot.Provider.Polly/PollyPoliciesDelegatingHandler.cs
Comment (RaynaldM) [reply src/Ocelot.Provider.Polly/PollyPoliciesDelegatingHandler.cs:2]: This is due to our corporate coding rules, and therefore to my VS22 setting (system first, then microsoft, then the rest).

**Comment by raman-m** [comment]
@RaynaldM commented [on Nov 17](#issuecomment-1816593136):
> For `should_be_invalid_re_route_using_downstream_http_version` UT, I noticed that the message changes depending on the OS language.

Really? 🤣 Surprise from France!

**Comment by ggnaegi** [comment]
> For "should_be_invalid_re_route_using_downstream_http_version" UT, I noticed that the message changes depending on the OS language. image So, I have changed it, and I fix ThenTheErrorsContains method so that it makes "Contains" and not "Equal".

Funny, I remember having seen this one, I thought it was fixed 🤣

File: src/Ocelot.Provider.Polly/OcelotBuilderExtensions.cs
Comment (ggnaegi) [inline src/Ocelot.Provider.Polly/OcelotBuilderExtensions.cs:20]: @RaynaldM The issue is that - DI - ```Dictionary<Type, Func<Exception, Error>>``` is not the same as ```IDictionary<Type, Func<Exception, Error>>```. In ```Ocelot.Requester.HttpExceptionToErrorMapper``` ocelot is trying to get a service by calling 
```c#
_mappers = serviceProvider.GetService<Dictionary<Type, Func<Exception, Error>>>();
```
... It will never work as expected. I agree with you, IDictionary is the correct option, but then you will need to modify ```HttpExceptionToErrorMapper``` and fix a test case in ```Ocelot.UnitTests.Requester``` in ```Should_return_error_from_mapper```, casting the Dictionary to IDictionary, 
```c# 
_services.AddSingleton((IDictionary<Type, Func<Exception, Error>>) errorMapping);
```

**Review by ggnaegi — CHANGES_REQUESTED** [review]

File: src/Ocelot.Provider.Polly/PollyPoliciesDelegatingHandler.cs
Comment (ggnaegi) [reply src/Ocelot.Provider.Polly/PollyPoliciesDelegatingHandler.cs]: same here! 😹

File: src/Ocelot.Provider.Polly/OcelotBuilderExtensions.cs
Comment (raman-m) [reply src/Ocelot.Provider.Polly/OcelotBuilderExtensions.cs:20]: Why `IDictionary<TKey, TValue>`? Do we use methods & props of this interface?
Is it OK to use most abstract interfaces like simple `ICollection<T>` or even `IEnumerable<T>`?

Comment (RaynaldM) [reply src/Ocelot.Provider.Polly/OcelotBuilderExtensions.cs]: back to Dictionary

**Review by ggnaegi — APPROVED** [review]

**Review by ggnaegi — APPROVED** [review]

**Comment by raman-m** [comment]
@ggnaegi  Sorry for being late!
I see you cannot wait for release commits from **main** branch 😄 
OK! Going to deliver release branch commits to develop...

**Comment by raman-m** [comment]
@ggnaegi  Watch for #1798 merging please to rebase current feature branch

**Comment by ggnaegi** [comment]
it's not me, @RaynaldM merged it 😸

**Comment by raman-m** [comment]
Ups, my fault! This PR is Raynald's one! 🤣 
So, Ray should approve and rebase this feature branch... Id he won't, then I will help in Nov'23 )))

**Comment by RaynaldM** [comment]
> Ups, my fault! This PR is Raynald's one! 🤣 Actually #1798 is merged. So, Ray should rebase this feature branch... If he won't, then I will help in Nov'23 )))

merge done

**Review by raman-m — APPROVED** [review]
Ready for delivery!

**Comment by raman-m** [comment]
QoS docs?
https://ocelot.readthedocs.io/en/latest/features/qualityofservice.html
It seems PR doesn't affect docs...

**Comment by raman-m** [comment]
@RaynaldM If you don't mind, I delete the feature branch...

**Pull request merged**
